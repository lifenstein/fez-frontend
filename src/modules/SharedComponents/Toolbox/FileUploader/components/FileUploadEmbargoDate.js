import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { DatePicker } from '@mui/x-date-pickers/DatePicker';
import { GENERIC_DATE_FORMAT } from 'config/general';
import TextField from '@mui/material/TextField';
import { AdapterMoment } from '@mui/x-date-pickers/AdapterMoment';
import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider';
import { red } from '@mui/material/colors';

export class FileUploadEmbargoDate extends PureComponent {
    static propTypes = {
        disabled: PropTypes.bool,
        minDate: PropTypes.instanceOf(Date),
        onChange: PropTypes.func,
        onKeyUp: PropTypes.func,
        value: PropTypes.string,
        fileUploadEmbargoDateId: PropTypes.string,
    };

    _onChange = value => {
        /* istanbul ignore else */
        if (this.props.onChange) this.props.onChange(value);
    };

    render() {
        const inputProps = {
            disableUnderline: true,
            'data-analyticsid': `${this.props.fileUploadEmbargoDateId}-input`,
            'data-testid': `${this.props.fileUploadEmbargoDateId}-input`,
        };

        return (
            <LocalizationProvider dateAdapter={AdapterMoment}>
                <DatePicker
                    value={this.props.value ? new Date(this.props.value) : null}
                    inputFormat={GENERIC_DATE_FORMAT}
                    minDate={this.props.minDate}
                    onChange={this._onChange}
                    disabled={this.props.disabled}
                    InputProps={inputProps}
                    renderInput={params => (
                        <TextField
                            {...params}
                            onKeyUpCapture={this.props.onKeyUp}
                            variant="standard"
                            sx={{
                                '& .MuiInput-input': {
                                    fontSize: 14,
                                    fontWeight: 400,
                                    ...(params.error ? { color: red[700] } : {}),
                                },
                            }}
                        />
                    )}
                />
            </LocalizationProvider>
        );
    }
}

export default FileUploadEmbargoDate;
