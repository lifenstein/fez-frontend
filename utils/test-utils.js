/* eslint-disable react/prop-types */
import React from 'react';
import PropTypes from 'prop-types';
import { render } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { Route } from 'react-router';
import { mui1theme } from 'config/theme';
import { Provider } from 'react-redux';
import { ThemeProvider as MuiThemeProvider } from '@material-ui/core/styles';
import { MuiPickersUtilsProvider } from '@material-ui/pickers';
import MomentUtils from '@date-io/moment';

import { getStore } from '../src/config/store';
import Immutable from 'immutable';
import { createMemoryHistory } from 'history';

const domTestingLib = require('@testing-library/dom');
const reactTestingLib = require('@testing-library/react');
const mime = require('mime-types');

const { configure } = domTestingLib;

configure(config => ({
    ...config,
    testIdAttribute: 'id',
}));

export const AllTheProviders = props => {
    return (
        <MuiThemeProvider theme={mui1theme}>
            <MuiPickersUtilsProvider utils={MomentUtils}>{props.children}</MuiPickersUtilsProvider>
        </MuiThemeProvider>
    );
};

AllTheProviders.propTypes = {
    children: PropTypes.node,
};

export const rtlRender = (ui, options) => render(ui, { wrapper: AllTheProviders, ...options });

export const renderWithRouter = (
    ui,
    { route = '/', history = createMemoryHistory({ initialEntries: [route] }), renderMethod = render } = {},
) => {
    return renderMethod(
        <AllTheProviders>
            <Router history={history}>{ui}</Router>
        </AllTheProviders>,
    );
};

export const WithRouter = ({ children, route = '/', history = createMemoryHistory({ initialEntries: [route] }) }) => (
    <AllTheProviders>
        <Router history={history}>{children}</Router>
    </AllTheProviders>
);
export const renderWithRedux = ({ initialState }) => render => {
    return {
        ...render,
        store: getStore({ initialState, history: render.history }),
    };
};

export const withRouter = ({
    route = '/',
    path = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
} = {}) => WrappedComponent => {
    return (
        <Router history={history}>
            <Route path={path} children={WrappedComponent} />
        </Router>
    );
};

export const withRedux = (initialState = Immutable.Map()) => WrappedComponent => {
    return <Provider store={getStore(initialState)}>{WrappedComponent}</Provider>;
};

export const WithReduxStore = ({ initialState = Immutable.Map(), children }) => (
    <Provider store={getStore(initialState)}>
        <AllTheProviders>{children}</AllTheProviders>
    </Provider>
);

export const defaultDatastream = {
    dsi_embargo_date: null,
    dsi_open_access: 1,
    dsi_label: '',
    dsi_copyright: null,
    dsi_state: 'A',
    dsi_size: 27932352,
};
export const createFezDatastreamInfoArray = (datastreams, pid = null, withPreview = true) => {
    let processed = datastreams.filter(datastream => typeof datastream === 'object' || typeof datastream === 'string');
    if (withPreview) {
        processed = [];
        datastreams.forEach(function(datastream) {
            processed.push(datastream);
            const filename = typeof datastream === 'object' ? datastream.dsi_dsid : datastream;
            const mimetype = mime.lookup(filename);
            // bail in case it's not derivable
            if (
                !filename.includes('.pdf') &&
                !mimetype.includes('image') &&
                !mimetype.includes('audio') &&
                !mimetype.includes('video')
            ) {
                return;
            }
            // eslint-disable-next-line no-nested-ternary
            const derivativeMimetype = mimetype.includes('audio')
                ? 'audio/mp3'
                : mimetype.includes('video')
                ? 'video/mp4'
                : 'image/jpg';

            (derivativeMimetype.includes('image') ? ['preview_', 'thumbnail_', 'web_'] : ['']).forEach(function(
                prefix,
            ) {
                processed.push({
                    dsi_dsid: `${prefix}${filename.split('.')[0]}_t.${derivativeMimetype.split('/').pop()}`,
                    dsi_mimetype: derivativeMimetype,
                });
            });
        });
    }

    return processed.map((datastream, index) => ({
        ...defaultDatastream,
        ...{
            dsi_pid: pid,
            dsi_order: index + 1,
            dsi_mimetype: mime.lookup(typeof datastream === 'object' ? datastream.dsi_dsid : datastream),
        },
        ...(typeof datastream === 'object'
            ? datastream
            : {
                  dsi_dsid: datastream,
              }),
    }));
};

export const withDatastreams = (sources, datastreams, callback) =>
    datastreams.map(item => {
        const source = sources.find(datastream =>
            datastream.dsi_dsid.includes(
                item.dsi_dsid
                    .replace('_t.', '.')
                    .replace('preview_', '')
                    .replace('thumbnail_', '')
                    .replace('web_', '')
                    .split('.')[0],
            ),
        );
        const isDerivative = item.dsi_dsid !== source.dsi_dsid;
        callback(item, source, isDerivative);
    });

export const getDatastreamByFilename = (filename, derivatives) =>
    derivatives.find(derivative => derivative.dsi_dsid === filename);

module.exports = {
    ...domTestingLib,
    ...reactTestingLib,
    rtlRender,
    renderWithRouter,
    renderWithRedux,
    withRedux,
    withRouter,
    AllTheProviders,
    WithReduxStore,
    WithRouter,
    createFezDatastreamInfoArray,
    getDatastreamByFilename,
    withDatastreams,
};
