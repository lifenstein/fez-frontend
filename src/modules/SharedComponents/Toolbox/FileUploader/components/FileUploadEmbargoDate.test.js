import React from 'react';
import FileUploadEmbargoDate from './FileUploadEmbargoDate';
import { rtlRender, fireEvent } from 'test-utils';

function setup(testProps = {}) {
    const props = {
        minDate: new Date('2016'),
        classes: {
            input: '',
        },
        ...testProps,
    };

    return rtlRender(<FileUploadEmbargoDate {...props} />);
}

describe('Component FileUploadEmbargoDate', () => {
    it('should render with default setup', () => {
        const { container } = setup({ value: '2016' });
        expect(container).toMatchSnapshot();
    });

    it('should render with no supplied date', () => {
        const { container } = setup({
            minDate: new Date('2016'),
            classes: {
                input: '',
            },
        });
        expect(container).toMatchSnapshot();
    });

    it('should render disabled', () => {
        const { container } = setup({ disabled: true, value: '2016' });
        expect(container).toMatchSnapshot();
    });

    it('should set correct date on date changed', () => {
        const onDateChangedTestFn = jest.fn();
        const props = {
            locale: {
                datePickerLocale: 'en-AU',
            },
            defaultConfig: {
                fileMetaKey: 'date',
                dateTimeFormat: global.Intl.DateTimeFormat,
                fieldName: 'accessDate',
            },
            onChange: onDateChangedTestFn,
            value: '2016',
        };

        const { container, getByRole } = setup(props);
        expect(container).toMatchSnapshot();

        fireEvent.change(getByRole('textbox'), { target: { value: '01/01/2018' } });
        expect(container).toMatchSnapshot();
        expect(onDateChangedTestFn).toHaveBeenCalled();
    });

    test('should style text field correctly for an invalid date', () => {
        const props = {
            fileUploadEmbargoDateId: 'component',
            value: '1111',
        };
        const { getByTestId } = setup(props);

        const textField = getByTestId('component-input'); // Add a data-testid to your TextField component

        // Get the computed style of the text field and check if the color property is red[700]
        const computedStyle = window.getComputedStyle(textField);
        expect(computedStyle.color).toBe('rgba(0, 0, 0, 0.87)');
    });

    it('should display the clear field', () => {
        const { container } = setup({ canBeCleared: true, value: '2016' });
        expect(container).toMatchSnapshot();
    });
});
