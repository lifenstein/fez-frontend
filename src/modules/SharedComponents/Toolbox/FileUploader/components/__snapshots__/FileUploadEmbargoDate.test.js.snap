// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component FileUploadEmbargoDate should display the clear field 1`] = `
<PickerWithState
  InputProps={
    Object {
      "classes": Object {
        "root": "",
      },
      "disableUnderline": true,
    }
  }
  allowKeyboardControl={true}
  autoOk={true}
  data-testid="undefined-input"
  format="DD/MM/YYYY"
  invalidDateMessage="Invalid Date Format"
  maxDate={2100-01-01T00:00:00.000Z}
  maxDateMessage="Date should not be after maximal date"
  minDate={2016-01-01T00:00:00.000Z}
  minDateMessage="Date should not be before minimal date"
  onChange={[Function]}
  openTo="date"
  value={2016-01-01T00:00:00.000Z}
  variant="inline"
  views={
    Array [
      "year",
      "date",
    ]
  }
/>
`;

exports[`Component FileUploadEmbargoDate should render disabled 1`] = `
<PickerWithState
  InputProps={
    Object {
      "classes": Object {
        "root": "",
      },
      "disableUnderline": true,
    }
  }
  allowKeyboardControl={true}
  autoOk={true}
  data-testid="undefined-input"
  disabled={true}
  format="DD/MM/YYYY"
  invalidDateMessage="Invalid Date Format"
  maxDate={2100-01-01T00:00:00.000Z}
  maxDateMessage="Date should not be after maximal date"
  minDate={2016-01-01T00:00:00.000Z}
  minDateMessage="Date should not be before minimal date"
  onChange={[Function]}
  openTo="date"
  value={2016-01-01T00:00:00.000Z}
  variant="inline"
  views={
    Array [
      "year",
      "date",
    ]
  }
/>
`;

exports[`Component FileUploadEmbargoDate should render with default setup 1`] = `
<PickerWithState
  InputProps={
    Object {
      "classes": Object {
        "root": "",
      },
      "disableUnderline": true,
    }
  }
  allowKeyboardControl={true}
  autoOk={true}
  data-testid="undefined-input"
  format="DD/MM/YYYY"
  invalidDateMessage="Invalid Date Format"
  maxDate={2100-01-01T00:00:00.000Z}
  maxDateMessage="Date should not be after maximal date"
  minDate={2016-01-01T00:00:00.000Z}
  minDateMessage="Date should not be before minimal date"
  onChange={[Function]}
  openTo="date"
  value={2016-01-01T00:00:00.000Z}
  variant="inline"
  views={
    Array [
      "year",
      "date",
    ]
  }
/>
`;

exports[`Component FileUploadEmbargoDate should render with default setup with styles 1`] = `
<FileUploadEmbargoDate
  classes={
    Object {
      "input": "FileUploadEmbargoDate-input-1",
    }
  }
  minDate={2016-01-01T00:00:00.000Z}
  value="2016"
/>
`;

exports[`Component FileUploadEmbargoDate should render with no supplied date 1`] = `
<PickerWithState
  InputProps={
    Object {
      "classes": Object {
        "root": "",
      },
      "disableUnderline": true,
    }
  }
  allowKeyboardControl={true}
  autoOk={true}
  data-testid="undefined-input"
  format="DD/MM/YYYY"
  invalidDateMessage="Invalid Date Format"
  maxDate={2100-01-01T00:00:00.000Z}
  maxDateMessage="Date should not be after maximal date"
  minDate={2016-01-01T00:00:00.000Z}
  minDateMessage="Date should not be before minimal date"
  onChange={[Function]}
  openTo="date"
  value={null}
  variant="inline"
  views={
    Array [
      "year",
      "date",
    ]
  }
/>
`;

exports[`Component FileUploadEmbargoDate should set correct date on date changed 1`] = `
<PickerWithState
  InputProps={
    Object {
      "classes": Object {
        "root": "",
      },
      "disableUnderline": true,
    }
  }
  allowKeyboardControl={true}
  autoOk={true}
  data-testid="undefined-input"
  format="DD/MM/YYYY"
  invalidDateMessage="Invalid Date Format"
  maxDate={2100-01-01T00:00:00.000Z}
  maxDateMessage="Date should not be after maximal date"
  minDate={2016-01-01T00:00:00.000Z}
  minDateMessage="Date should not be before minimal date"
  onChange={[Function]}
  openTo="date"
  value={2016-01-01T00:00:00.000Z}
  variant="inline"
  views={
    Array [
      "year",
      "date",
    ]
  }
/>
`;

exports[`Component FileUploadEmbargoDate should set correct date on date changed 2`] = `
<PickerWithState
  InputProps={
    Object {
      "classes": Object {
        "root": "",
      },
      "disableUnderline": true,
    }
  }
  allowKeyboardControl={true}
  autoOk={true}
  data-testid="undefined-input"
  format="DD/MM/YYYY"
  invalidDateMessage="Invalid Date Format"
  maxDate={2100-01-01T00:00:00.000Z}
  maxDateMessage="Date should not be after maximal date"
  minDate={2016-01-01T00:00:00.000Z}
  minDateMessage="Date should not be before minimal date"
  onChange={[Function]}
  openTo="date"
  value={2016-01-01T00:00:00.000Z}
  variant="inline"
  views={
    Array [
      "year",
      "date",
    ]
  }
/>
`;
